File: s2core/src/main/scala/org/apache/s2graph/core/tinkerpop/structure/S2VertexProperty.scala
//package org.apache.s2graph.core.tinkerpop.structure
//
//import org.apache.tinkerpop.gremlin.structure.util.ElementHelper
//import org.apache.tinkerpop.gremlin.structure.{Property, Vertex, VertexProperty}
//import scala.collection.JavaConversions._
//import java.util
//
//class S2VertexProperty[V](val id: AnyRef,
//                          val vertex: S2Vertex,
//                          val key: String,
//                          val value: V,
//                          val kvs: AnyRef*) extends VertexProperty[V] {
//
//  val properties: scala.collection.mutable.Map[String, Property[_]] = scala.collection.mutable.Map.empty
//
//  ElementHelper.legalPropertyKeyValueArray(kvs)
//  ElementHelper.attachProperties(vertex, kvs)
//
//  override def properties[U](keys: String*): util.Iterator[Property[U]] = {
//    val set = for {
//      key <- keys
//      prop = property[U](key) if prop != null
//    } yield {
//      prop
//    }
//    set.iterator
//  }
//
//  override def element(): Vertex = vertex
//
//  override def isPresent: Boolean = true
//
//  override def property[U](key: String, v: U): Property[U] = {
//    val newProperty = new S2Property[U](vertex, key, v)
//    properties.put(key, newProperty)
//    newProperty
//  }
//
//  override def remove(): Unit = ???
//
//}

File: s2core/src/main/java/org/apache/s2graph/core/tinkerpop/structure/S2VertexProperty.java
package org.apache.s2graph.core.tinkerpop.structure;

import org.apache.tinkerpop.gremlin.structure.Property;
import org.apache.tinkerpop.gremlin.structure.Vertex;
import org.apache.tinkerpop.gremlin.structure.VertexProperty;

import java.util.*;

public class S2VertexProperty<V> implements VertexProperty<V> {
    private S2Vertex element;
    private String key;
    private V value;
    private Map<String, Property<?>> props;

    public S2VertexProperty(S2Vertex element, String key, V value) {
        this.element = element;
        this.key = key;
        this.value = value;
        this.props = new HashMap<>();
    }

    @Override
    public String key() {
        return key;
    }

    @Override
    public V value() throws NoSuchElementException {
        return value;
    }

    @Override
    public boolean isPresent() {
        return true;
    }

    @Override
    public Vertex element() {
        return element;
    }

    @Override
    public void remove() {

    }

    @Override
    public Object id() {
        S2VertexId vertexId = element.getVertexId();
        return vertexId.getServiceName() + "|" +
                vertexId.getColumnName() + "|" +
                vertexId.getId() + "|" + key;
    }

    @Override
    public <V> Property<V> property(String s, V v) {
        VertexProperty<V> newProperty = new S2VertexProperty<V>(element, key, v);
        props.put(key, newProperty);
        return newProperty;
    }

    @Override
    public <U> Iterator<Property<U>> properties(String... propertyKeys) {
        List<Property<U>> ls = new ArrayList<>();
        for (String propertyKey : propertyKeys) {
            ls.add(property(propertyKey));
        }
        return ls.iterator();
    }
}
